### PR TITLE
Multiaddress refactor for issue #20

### PR DESCRIPTION
**Pull request by JuanuMusic**
Replaced `Multiaddr` with `Multiformats.Multiaddress` project.
Chat sample working.
All tests passing.

SOLVES: https://github.com/NethermindEth/dotnet-libp2p/issues/20

### CODE REVIEW

**Comment by flcl42** [comment]
Thanks a lot for the contribution! Looks great!

Could you integrate cs-multiaddress as a git submodule?
Could you remove that old implementation please? I mean [MultiAddr](https://github.com/JuanuMusic/dotnet-libp2p/blob/multiaddress-refactor/src/libp2p/Libp2p.Core/Multiaddr.cs)

**Comment by JuanuMusic** [comment]
Makes sense.
I have some updates done on the cs-multiaddress but I think I could
implement them as extensions.

Will do!

**Comment by JuanuMusic** [comment]
Lets hold off on Merging. I found a couple issues running PubsubChat.

**Comment by flcl42** [comment]
> Lets hold off on Merging. I found a couple issues running PubsubChat.
> 
> @flcl42 @rubo @chertby does the `Libp2p.Protocols.Quic` protocol support `QUIC` as well as `QUICv1` ?

`quic` is obsolete, there are no plans to support it and it has never been tested.

**Comment by JuanuMusic** [comment]
Understood. Im gettng errors actuallly by .net7 not having support for it.
Ill push the PR then and we can move forward from there.

File: src/libp2p/Libp2p/MultiaddressBasedSelectorProtocol.cs [hunk 1]
```diff
@@ -20,11 +20,15 @@ public class MultiaddressBasedSelectorProtocol(ILoggerFactory? loggerFactory = n
...
-        // TODO: Validate support for BOTH or just quicv1 and deprecate quic
-        if (context.LocalPeer.Address.Has<QUIC>()|| context.LocalPeer.Address.Has<QUICv1>())
+        // TODO: deprecate quic
+        if (context.LocalPeer.Address.Has<QUIC>())
         {
             protocol = channelFactory!.SubProtocols.FirstOrDefault(proto => proto.Id.Contains("quic")) ?? throw new ApplicationException("QUIC is not supported");
         }
+        else if (context.LocalPeer.Address.Has<QUICv1>())
+        {
+            protocol = channelFactory!.SubProtocols.FirstOrDefault(proto => proto.Id.Contains("quic-v1")) ?? throw new ApplicationException("QUICv1 is not supported");
```
Comment (flcl42) [inline src/libp2p/Libp2p/MultiaddressBasedSelectorProtocol.cs:30]: `QUIC` should be declined without an attempts to search for it.

**Comment by flcl42** [comment]
> Understood. Im gettng errors actuallly by .net7 not having support for it. Ill push the PR then and we can move forward from there.

It may work, but you need to customize your .NET: https://github.com/NethermindEth/dotnet-libp2p/blob/main/src/libp2p/Libp2p.Protocols.Quic/README.md. It's a pity that we need such manipulations, but for now it's the only way and known .NET issue.

**Comment by JuanuMusic** [comment]
@flcl42 Looks like the tests failed because of a Server error on nuget. Might have been a temporary issue on their servers?
I am running it and it works perfectly.
Can you trigger the checks again?
image

**Review by flcl42 — APPROVED** [review]
Looks good. We can make submodule from cs-multiaddress in a separate pr

**Comment by JuanuMusic** [comment]
Hmmm I added it as a git submodule, but looks like its just taking it as files from the same repo.
Do you need me to do anything or this is good to be merged?

**Comment by flcl42** [comment]
Merged! Thanks a lot!

**Pull request merged**